### ISSUE TITLE
Refactor status-go geth packages

### ISSUE THREAD

**Issue by adambabik** [status-im/ideas#63]
## Preamble

    Idea: DEV63
    Title: Refactor status-go node package
    Status: Draft
    Created: 2017-12-20

## Summary
`geth/` packages require refactoring as they are tightly coupled, hard to maintain and extend. In this idea, we would like to fix all these issues and make status-go architecture easier to work with for new core members and external contributors.

## Swarm Participants
- Lead Contributor: @adambabik
- Contributors: 
  - @dshulyak (20h/week)
  - @adriacidre
- Tester/Evaluator: TBD

## Product Overview
We have a bunch of problems related to the `geth/` packages:
* `NodeManager` contains a lot of complexity, especially related to starting and stopping a node,
* `geth/account` unnecessarily depends on packages like `geth/rpc`,
* TBD

In this idea, our goal is to come up with a better architecture design for these packages, removing unnecessary dependencies and removing huge interfaces that make the code less clear.

### Product Description
It does not add any new features from the user perspective but it's a part of making status-go better tested, easier to reason about and extend, also by external contributors.

### Requirements & Dependencies
All existing tests should pass. If we need to change the interface in `library.go`, we need to communicate that and work with status-react team to figure out the best way to do it.

### Minimum Viable Product
Goal Date: 2018-02-02

Description: Minimal implementation of `Node` that can be started and stopped and an RPC client can be connected to it. It should be preceded with code examples how the new interface of `Node` will look like. 

### Iteration 1
Goal Date: 2018-02-09

Description: The rest of the `Node` features that are currently implemented and required by other components.

### Iteration 2 : [AccountManager cleanup](https://github.com/status-im/status-go/pull/753)
Goal Date: 2018-03-23

Description: 
AccountManager shouldn't know nothing about whisper or Nodemanager, so on this iteration we will remove this relations

### Iteration 3 : [TransactionManager cleanup](https://github.com/status-im/status-go/issues/772)
Goal Date: 2018-04-01

Description: 
TransactionManager shouldn't be directly depending on AccountManager and NodeManager 

### Iteration 4 : [common.Nodemanager removal](https://github.com/status-im/status-go/issues/779)
Goal Date: TBD

Description:
common.NodeManager interface and all its mocks have to be removed from status-go 

### Iteration 5 : common package removal? _(*Maybe is still to big, try to split it)_
Goal Date: TBD

Description:
TBD

### Iteration 6 : NodeManager refactor...
Goal Date: TBD

Description:
TBD

## Success Metrics
Test metrics like code coverage for the new package should be around 80%. There should be no race conditions.

## Supporting Role Communication
In Iteration 3, we will need help from the status-react or Q&A team with building the app.

## Copyright
Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

**Comment by dshulyak** [comment]
I pledge 20 hours a week. I will start by addressing several pain points:

- simplify API for managing status node
- reduce code coupling, mainly by removing usage of nodeManager where it is not necessary. for example transaction manager needs couple of values from config and rpc client. and there is lots of places like this

progress will be tracked as a part of https://github.com/status-im/status-go/projects/7

**Comment by adambabik** [comment]
MVP, Iteration 1 and Iteration 2 are finished after #584 got merged.

**Comment by adambabik** [comment]
Update: we are waiting for the current release to be finished and published before we create another PR with status-go update. We don't want to push more changes into the current release.

**Comment by oskarth** [comment]
Replaced by https://github.com/status-im/ideas/blob/master/ideas/63-refactor-geth-packages.md
